docs/notebooks/burger.ipynb

```diff
@@ -0,0 +1,788 @@
+    "\n",
+    "def convergence_rates(E_values, eps_values, show=True):\n",
+    "    r = []\n",
+    "    for i in range(1, len(eps_values)):\n",
+    "        r.append(np.log(E_values[i] / E_values[i - 1])\n",
+    "                 / np.log(eps_values[i] / eps_values[i - 1]))\n",
+    "    if show:\n",
+    "        print(\"Computed convergence rates: {}\".format(r))\n",
+    "    return r"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We start by computing the gradient without any checkpointing. Then, we compute the gradient using the `SingleMemoryStorageSchedule`, `SingleMemoryStorageSchedule`, `Revolve` and `HRevolve` checkpointing schedules. We then compare the gradients obtained from each checkpointing schedule to the gradient obtained without checkpointing. We expect the gradients to be the same. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Computed residuals: [1.0338017632341477e-07]\n",
+      "Computed residuals: [1.0338017632341477e-07, 2.5846557596146064e-08]\n",
+      "Computed residuals: [1.0338017632341477e-07, 2.5846557596146064e-08, 6.4618285621002554e-09]\n",
+      "Computed residuals: [1.0338017632341477e-07, 2.5846557596146064e-08, 6.4618285621002554e-09, 1.6154808828890686e-09]\n",
+      "Computed convergence rates: [1.9999155166047717, 1.9999577660400731, 1.999978796875359]\n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "[1.9999155166047717, 1.9999577660400731, 1.999978796875359]"
+      ]
+     },
+     "execution_count": 5,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "burger = BurgersEquation(model, u0, mesh)  # Create the burger's equation object.\n",
+    "burger.forward(0, model[\"max_n\"], storage=StorageType.WORK, write_adj_deps=True)  # Solve the forward problem.\n",
+    "burger.adjoint(0, model[\"max_n\"], clear_adj_deps=True)  # Solve the adjoint problem.\n",
+    "base_gradient = burger.gradient()  # Compute the gradient.\n",
+    "h = np.random.rand(model[\"nx\"])  # Random perturbation.\n",
+    "functional = burger.functional()\n",
+    "E, epsilon = taylor_remainder_test(burger.gradient(), functional, u0, h)  # Taylor remainder test.\n",
+    "convergence_rates(E, epsilon)  # Compute the convergence rates."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We now compute the adjoint-based gradient using `SingleMemoryStorageSchedule` checkpointing approach. The `SingleMemoryStorageSchedule` stores the forward data of each time-step in working memory. As explained in the [notebook with illustrative example](https://nbviewer.org/github/firedrakeproject/checkpoint_schedules/blob/main/docs/notebooks/tutorial.ipynb), this schedule does not require the maximal step (`model[\"max_n\"]`)."
```

Comment:
This is conceptually the same as the previous case -- except that a scheduler is used to implement it.

Reply:
I removed this code cell.